src/redisearch_rs/rlookup/src/lookup.rs

```diff
@@ -239,6 +266,392 @@ impl<'a> RLookupKey<'a> {
+        mem::replace(self.next.get_mut(), next)
+    }
+
+    fn assert_valid(&self, tail: &Self, ctx: &str) {
+        assert!(
+            !self.flags.intersects(TRANSIENT_FLAGS),
+            "{ctx}key flags must not contain transient ({TRANSIENT_FLAGS:?}) flags. Found {:?}.",
+            self.flags
+        );
+        assert!(
+            ptr::eq(self.name, self._name.as_ptr()),
+            "{ctx}`key.name` did not match `key._name`. ({self:?})",
+        );
+        if let Some(path) = self._path.as_ref() {
+            assert!(
+                ptr::eq(self.path, path.as_ptr()),
+                "{ctx}`key._path` is present, but `key.path` did not match `key._path`. ({self:?})"
+            );
+        } else {
+            assert!(
+                ptr::eq(self.path, self._name.as_ptr()),
+                "{ctx}`key._path` is not present, but `key.path` did not match `key._name`. ({self:?})"
+            );
+        }
+        assert_eq!(
+            self.name_len,
+            self._name.count_bytes(),
+            "{ctx}`key.name_len` did not match `key._name` length"
+        );
+
+        if ptr::eq(self, tail) {
+            assert_eq!(
+                self.next(),
+                None,
+                "{ctx}tail key must not have a next link; node={self:#?}",
+            );
+        }
+
+        if let Some(next) = self.next() {
+            assert_ne!(
+                // Safety:
+                NonNull::from(unsafe { &next.as_ref().next }),
+                NonNull::from(&self.next),
+                "{ctx}key's next link cannot be to itself; node={self:#?}",
+            );
+        }
+    }
+}
+
+// ===== impl KeyList =====
+
+#[cfg_attr(not(test), expect(unused, reason = "used by later stacked PRs"))]
+impl<'a> KeyList<'a> {
+    /// Construct a new, empty `KeyList`.
+    pub const fn new() -> Self {
+        Self {
+            head: None,
+            tail: None,
+            rowlen: 0,
+        }
+    }
+
+    /// Insert a `RLookupKey` into this `KeyList` and return a mutable reference to it.
+    ///
+    /// The key will be owned by the list and freed when dropping the list.
+    fn push(&mut self, mut key: RLookupKey<'a>) -> Pin<&mut RLookupKey<'a>> {
+        #[cfg(debug_assertions)]
+        self.assert_valid("KeyList::push before");
+
+        key.dstidx = u16::try_from(self.rowlen).expect("conversion from u32 RLookup::rowlen to u16 RLookupRow::dstidx overflowed. This is a bug!");
+
+        // Safety: RLookup never hands out mutable references to the key (except `Pin<&mut T>` which is fine)
+        // and never copies, or memmoves the memory internally.
+        let mut ptr = unsafe { RLookupKey::into_ptr(Box::pin(key)) };
+
+        if let Some(mut tail) = self.tail.take() {
+            // if we have a tail we also must have a head
+            debug_assert!(self.head.is_some());
+
+            // Safety: We know we can borrow tail here, since we mutably borrow the RLookup
```

Comment:
We mutably borrow KeyList or RLookup?

Reply:
Good catch, leftover from a refactor. It's borrowing the `KeyList` of course. Changed 👍🏻